iothub/device/src/Pipeline/RetryDelegatingHandler.cs

```diff
@@ -1090,12 +1091,41 @@ protected override void Dispose(bool disposing)
                     {
                         _handleDisconnectCts?.Cancel();
                         _handleDisconnectCts?.Dispose();
-                        if (_handlerSemaphore != null && _handlerSemaphore.CurrentCount == 0)
+
+                        if (_clientOpenCloseSemaphore != null && _clientOpenCloseSemaphore.CurrentCount == 0)
+                        {
+                            _clientOpenCloseSemaphore.Release();
```

Comment:
We've seen timing issues recently where we get an error here. I'd say wrap in a catch, maybe?

Reply:
This release is attempted only if the above expression indicates that the semaphore is currently held by a thread. This shouldn't give us any error. What do you think?

Reply:
I'm saying we've seen errors in our pipeline that it couldn't release because it would exceed the count. Meaning the code sections above that had the semaphore were still in it, our if statement confirmed they still had it, we tried to release it for them (suspect anyway), but before we could they released it.

I'm thinking we should either not release it here (just dispose), or if we do, we should do each of them inside a try/catch.

Reply:
I can add some additional checks around this cleanup code so that failures here don't crash the application, sure.